compiler/lib/lambda_lifting.ml

```diff
@@ -0,0 +1,227 @@
+           deeply nested closures. *)
+        let does_not_start_with_closure l =
+          match l with
+          | Let (_, Closure _) :: _ -> false
+          | _ -> true
+        in
+        let rec rewrite_body first st l =
+          match l with
+          | (Let (f, (Closure (_, (pc', _)) as cl)) as i) :: rem
+            when first && does_not_start_with_closure rem ->
+              let program, functions =
+                traverse var_depth st pc' (depth + 1) (depth + threshold)
+              in
+              if compute_depth program pc' + 1 >= threshold
+              then (
+                let free_vars = collect_free_vars program var_depth (depth + 1) pc' in
+                let s =
+                  Var.Set.fold
+                    (fun x m -> Var.Map.add x (Var.fork x) m)
+                    free_vars
+                    Var.Map.empty
+                in
+                let program = Subst.cont (Subst.from_map s) pc' program in
+                let f' = try Var.Map.find f s with Not_found -> Var.fork f in
+                let s = Var.Map.bindings (Var.Map.remove f s) in
+                let f'' = Var.fork f in
+                if debug ()
+                then
+                  Format.eprintf
+                    "LIFT %s (depth:%d free_vars:%d inner_depth:%d)@."
+                    (Code.Var.to_string f'')
+                    depth
+                    (Var.Set.cardinal free_vars)
+                    (compute_depth program pc');
+                let pc'' = program.free_pc in
+                let bl = { params = []; body = [ Let (f', cl) ]; branch = Return f' } in
+                let program =
+                  { program with
+                    free_pc = pc'' + 1
+                  ; blocks = Addr.Map.add pc'' bl program.blocks
+                  }
+                in
+                let functions =
+                  Let (f'', Closure (List.map snd s, (pc'', []))) :: functions
+                in
+                let rem', st = rewrite_body false (program, functions) rem in
+                Let (f, Apply (f'', List.map fst s, true)) :: rem', st)
+              else
+                let rem', st = rewrite_body false (program, functions) rem in
+                i :: rem', st
+          | (Let (_, Closure (_, (pc', _))) as i) :: rem ->
+              let st = traverse var_depth st pc' (depth + 1) limit in
+              let rem', st = rewrite_body false st rem in
+              i :: rem', st
+          | i :: rem ->
+              let rem', st = rewrite_body (does_not_start_with_closure l) st rem in
+              i :: rem', st
+          | [] -> [], st
+        in
+        let body, (program, functions) =
+          rewrite_body true (program, functions) block.body
+        in
+        ( { program with blocks = Addr.Map.add pc { block with body } program.blocks }
+        , functions ))
+    pc
+    program.blocks
+    (program, functions)
+
+let f program =
```

Comment:
Why is this only applied with effect ? 

Maybe make `threshold` configurable (using `Config.Param`) and add some expect tests in  compiler/test-compiler (with a small threshold)

Reply:
We have never got any report of nested functions being an issue. While it is very much an issue after the CPS transform.

I will add a test.